ProjectSidewalk/SidewalkWebpage
# 1130 update mission infrastructure

Resolves #1130 
Resolves #646 
Resolves #1095 

It is finally here! The moment you've all been waiting for... More flexible mission code!!! The code base is 2% to 3% smaller, missions are more oriented towards the user instead of our back-end, you can query for the labels placed during a mission, and so much more!

Testing:
1. I would restore a second database dump in your local environment alongside your main one. I would get an older dump (that is smaller). I called my database `sidewalk-small`. Then edit the `conf/application.conf` file line 38 to have the different database name. I would do this because this update will make changes to the database that can't be undone without restoring the entire database from scratch again.
1. Load up the landing page and check that the tables in the database have been updated appropriately:
    1. The `mission_user` table should be gone
    1. The `mission` table should have completely different columns from the old version
    1. There should be a new `mission_type` table containing 4 entries
    1. The `label`, `audit_task_comment`, `audit_task_interaction`, `audit_task_environment`, and `audit_task_incomplete` tables should all have a new `mission_id` column (and all these tables will now be empty).
    1. The `audit_task_comment` table should now have an `audit_task_id` column.
1. Switch back to the `develop` branch and reload the landing page, it should undo all of those database changes listed above (except it will not fill in the tables that were emptied). You can then go back to the `1130-update-mission-infrastructure` branch.
1. Go through the tutorial and a few missions. I've touched pretty much every part of the interface in this PR, so the person testing needs to look really carefully and every single part of the UI to make sure that it makes sense. Here is a list of some things to check, but please do a sanity check on everything you can see.
    1. The "Current Neighborhood" section in the top-right of the audit page needs to update correctly while you audit, and have correct values when you refresh the page partially through a mission and/or street.
    1. The "Current Mission" section right below that should appropriate text at all times, including saying "Complete the onboarding tutorial!" during onboarding and having the correct mission distance at other times. The mission progress bar should increment as expected. It should get to approximately 100% at the end of the mission.
    1. The label count section below that should work as expected.
    1. The mission start modal (where it says what your next mission is and you click "Ok") should have appropriate text for every mission. So the first mission should have text referring to a "first mission", but only on the first mission. The distance that is given for the mission should be the same as it is shown everywhere else in the UI during the mission.
    1. All the numbers on the mission complete modal should look correct (label counts, distance in the mission, total distance audited in the neighborhood, etc).
1. Make sure that everything works as expected when you use the jump button (both in the beginning, middle, and end of a mission).
1. When auditing a new neighborhood, I have the default sequence of missions to be 500ft, 500ft, 1000ft, 1000ft, then all the remaining being 1/4 mile
1. Check that resuming a mission works correctly: partially audit a street, check the mission completion percentage, refresh the page, and your mission completion percentage should remain the same (it used to reset you back to the completion amount from the start of that street you partially did).
1. Check that the retake tutorial button works as expected.
1. You should be able to refresh the page at weird times without anything looking incorrect in the UI or in the `mission` table. Try refreshing the page mid-mission, on the mission start screen, on the mission complete screen, etc. Check for anything that seems out of place if you do that.
1. Complete a mission. Check the `audit_task_interaction` table for an entry with the action `MissionComplete`. It should now have the metadata `missionId`, `missionType`, `distanceMeters`, and `regionId` in the `note` field. It used to have `missionLabel`, `missionDistance`, and `neighborhoodId`.
1. Check that the "skip tutorial" option works correctly when you are shown the tutorial.
1. Check that there are the correct number of completed missions in the `mission` table. There should be a max of 1 incomplete mission per user per neighborhood.
1. Finish auditing an entire neighborhood (I usually do this by finding a really small one and then jumping through the streets, clicking the "I cannot go in the direction you want me to walk" option). The last mission in the neighborhood should be the exact distance left to audit in that neighborhood (although you will probably hit the "neighborhood complete" screen before you reach 100% completion of the mission, but I don't think that is super important). The page should refresh and you should be sent to a new neighborhood.
1. After completing that neighborhood, start to audit an adjacent neighborhood. In the old code, since neighborhoods shared streets, when you went to this second neighborhood you would be greeted by a mission complete modal and such, which didn't make sense. Instead, it should look like you haven't started auditing this neighborhood yet, except that when you finish a mission, you will notice that the total distance audited in that neighborhood should include the shared streets.
1. Test out starting your auditing by picking a neighborhood through one of the choropleths. Make sure the audit page loads correctly and nothing seems out of place. Make sure that you can resume a partially completed mission that way.
1. If you did the above checks as an anon user, do them again as a registered user, or vice versa.
1. Check that the turker-specific things work correctly
    1. Log in as a turker by opening a new incognito/private browser and using the URL `localhost:9000/?referrer=mturk&hitId=h1&workerId=testTurker1&assignmentId=a1`
    1. You should be sent directly to the audit page. Make sure that the "skip tutorial" is _not_ an option.
    1. Whenever you start a new mission you should also see "Reward on satisfactory completion: $X.XX" on the mission start modal. Make sure that the dollar amount is equal to the distance they are auditing in miles times $4.17 (which would be $0.39 for that first 500 ft mission, for example).
    1. You should see that same mission reward amount in the "Current Mission" section of the UI while the user is auditing, and the "Total Reward Earned" right below that number should also be correct (note that a turker is paid $0.43 for the tutorial). This amount should increase by the correct amount after every mission.
    1. When you finish a mission, the same reward amount for completing a mission should be shown each time.
    1. When you finish your first non-tutorial mission, you should see a confirmation code appear above the continue button. After clicking continue, that button should remain reachable from a button on the left above the sound button.
    1. After the 2nd mission, you should get a short survey that pops up.
    1. Try doing things to mess with the reward amounts by refreshing the page at weird times, entering the audit page through a choropleth, etc etc.
    1. Go to the user dashboard, and you should see the same "Total reward earned" amount, and it should be correct and the same as is shown on the audit page.
